File: apps/web/src/lib/plugins.ts
"use client";

/** Imported algorithm bundles, from the browser's side.
 *
 * The sibling of `models.ts`, and separate for the reason the API keeps
 * them separate: a model is weights for a controller the platform
 * already has, a bundle is a controller it has never seen. Sharing a
 * module would mean one `ModelSummary` with half its fields meaningless
 * depending on which kind of thing it described.
 *
 * Upload goes through `XMLHttpRequest` for the same reason it does
 * there — `fetch` still cannot report upload progress, and an upload
 * with no progress looks exactly like a hung page.
 */

import { API_BASE } from "./api";
import { authFetch, loadSession } from "./auth";
import type { ValidationStatus } from "./models";

/** May this revision be picked, and if not, is that final.
 *
 * A third value on the field that already answered the question rather
 * than a second field beside it. `held` is a reviewer pulling a revision
 * back while they look at something; `disabled` is terminal, and
 * terminal on purpose — "turn it back on" and "upload the fixed one"
 * should not both exist, because only the second is honest about what
 * changed in between.
 */
export type BundleStatus = "active" | "held" | "disabled";

export interface PluginBundleSummary {
  id: string;
  name: string;
  version: string;
  description: string;
  plugin_id: string;
  plugin_version: string;
  /** Which upload of this `plugin_id` this is. Publication names a
   * revision rather than a bundle, so this is what a history row and a
   * pinned run both point at. */
  revision: number;
  role: string;
  requirements: string[];
  robot_profile_id: string;
  original_filename: string;
  file_size: number;
  checksum: string;
  status: BundleStatus;
  validation_status: ValidationStatus;
  validation_message: string;
  owned: boolean;
  created_at: string;
  updated_at: string;
}

/** The host's own preflight answer, field for field.
 *
 * Rendered rather than reworded. Every entry here is a list the host
 * filled in, and `why` is its own one-line explanation — a UI that
 * summarised them into something friendlier would be inventing a
 * diagnosis the platform did not make.
 */
export interface HostCompatibility {
  state: string;
  runnable: boolean;
  evidence_class: string;
  runtime_lane: string;
  why: string;
  missing_capabilities: string[];
  missing_providers: string[];
  missing_runtime: string[];
  incompatible_action_types: string[];
  incompatible_dynamics: string[];
  incompatible_execution_models: string[];
  fairness_refusals: string[];
  undeclared_providers: string[];
  graph_problems: string[];
  provider_order: string[];
  oracle_providers: string[];
}

/** One act of putting a revision in front of everybody.
 *
 * Append-only, and the two closing dates are separate columns because
 * they are separate facts: `superseded_at` means a newer revision took
 * its place, `unpublished_at` means somebody withdrew it and nothing
 * replaced it. A single `ended_at` would answer "is it current?"
 * equally well and "why did it stop?" not at all, which is the question
 * a reader of the history is actually asking.
 */
export interface PluginPublication {
  bundle_id: string;
  revision: number;
  published_at: string;
  published_by_user_id: string | null;
  superseded_at: string | null;
  unpublished_at: string | null;
  reason: string;
  is_current: boolean;
}

export interface PluginEvent {
  sequence: number;
  revision: number;
  actor_user_id: string | null;
  actor_roles: string;
  authorized_capability: string;
  action: string;
  reason: string;
  created_at: string;
}

export interface PluginBundleDetail {
  bundle: PluginBundleSummary;
  compatibility: HostCompatibility;
  /** The revision an engineer would actually get, or `null` while
   * nobody has published one. Sent to every reader, not just a
   * reviewer: "why is this not in my picker?" is a question the person
   * picking has to be able to answer without asking one. */
  published_revision: number | null;
  /** The half that describes code rather than capability. Absent — not
   * empty — for a reader without `algorithm.inspect`, so a component
   * that finds them missing should draw nothing rather than an empty
   * box. */
  manifest?: Record<string, unknown> | null;
  entry_point?: string | null;
  publications?: PluginPublication[] | null;
}

export function pluginEvents(bundleId: string): Promise<PluginEvent[]> {
  return authFetch<PluginEvent[]>(`/algorithms/plugins/${bundleId}/events`);
}

/** The five governed acts, which differ only in the word in the path.
 *
 * All five answer 404 rather than 403 while
 * `PLANBENCH_ALGORITHM_GOVERNANCE` is off, because the truth is "this
 * deployment has not turned publishing on" rather than "you may not" —
 * and a caller that can tell the difference hides the button instead of
 * offering one that always fails.
 */
function govern(
  bundleId: string,
  act: string,
  reason: string,
): Promise<PluginBundleSummary> {
  return authFetch<PluginBundleSummary>(`/algorithms/plugins/${bundleId}/${act}`, {
    method: "POST",
    body: JSON.stringify({ reason }),
  });
}

export const publishPlugin = (id: string, reason = "") => govern(id, "publish", reason);
export const unpublishPlugin = (id: string, reason: string) => govern(id, "unpublish", reason);
export const holdPlugin = (id: string, reason: string) => govern(id, "hold", reason);
export const releasePluginHold = (id: string, reason = "") => govern(id, "release-hold", reason);
export const disablePlugin = (id: string, reason: string) => govern(id, "disable", reason);

export interface ImportFields {
  name: string;
  version: string;
  description: string;
  robotProfileId: string;
  bundleFile: File;
}

export function listPlugins(): Promise<PluginBundleSummary[]> {
  return authFetch<PluginBundleSummary[]>("/algorithms/plugins");
}

export function getPlugin(bundleId: string): Promise<PluginBundleDetail> {
  return authFetch<PluginBundleDetail>(`/algorithms/plugins/${bundleId}`);
}

export function setPluginStatus(
  bundleId: string,
  status: BundleStatus,
): Promise<PluginBundleSummary> {
  return authFetch<PluginBundleSummary>(`/algorithms/plugins/${bundleId}`, {
    method: "PATCH",
    body: JSON.stringify({ status }),
  });
}

/** Change what is *written about* a bundle, never what it is.
 *
 * The manifest and the archive are identity: a bundle whose code could
 * be edited in place would let one `plugin_id` at one version mean two
 * different controllers, and every result recorded against it would stop
 * being attributable. The server refuses those fields; this only ever
 * sends the four it accepts.
 */
export function updatePlugin(
  bundleId: string,
  changes: {
    name?: string;
    version?: string;
    description?: string;
    robot_profile_id?: string;
  },
): Promise<PluginBundleSummary> {
  return authFetch<PluginBundleSummary>(`/algorithms/plugins/${bundleId}`, {
    method: "PATCH",
    body: JSON.stringify(changes),
  });
}

export function revalidatePlugin(bundleId: string): Promise<PluginBundleSummary> {
  return authFetch<PluginBundleSummary>(`/algorithms/plugins/${bundleId}/validate`, {
    method: "POST",
  });
}

export function importPlugin(
  fields: ImportFields,
  onProgress: (percent: number) => void,
): Promise<PluginBundleSummary> {
  const session = loadSession();
  const body = new FormData();
  body.append("name", fields.name);
  body.append("version", fields.version || "1");
  body.append("description", fields.description);
  body.append("robot_profile_id", fields.robotProfileId);
  body.append("bundle", fields.bundleFile);

  return new Promise<PluginBundleSummary>((resolve, reject) => {
    const request = new XMLHttpRequest();
    request.open("POST", `${API_BASE}/api/v1/algorithms/plugins`);
    if (session) request.setRequestHeader("Authorization", `Bearer ${session.token}`);

    request.upload.onprogress = (event) => {
      if (event.lengthComputable) onProgress(Math.round((event.loaded / event.total) * 100));
    };
    request.onload = () => {
      let parsed: unknown = null;
      try {
        parsed = JSON.parse(request.responseText);
      } catch {
        // A non-JSON body means something upstream failed; the status is
        // then the only thing worth reporting.
      }
      if (request.status >= 200 && request.status < 300) {
        resolve(parsed as PluginBundleSummary);
        return;
      }
      const message =
        (parsed as { error?: { message?: string } })?.error?.message ??
        `Import failed (${request.status})`;
      reject(new Error(message));
    };
    request.onerror = () => reject(new Error("The upload could not reach the server."));
    request.onabort = () => reject(new Error("Upload cancelled."));
    request.send(body);
  });
}

/** Whether this bundle may be offered as a candidate.
 *
 * **Two conditions, and both are load-bearing.** `active` is a decision
 * somebody made; `loaded` is the conformance suite having actually run
 * the plugin and found it behaved. `structural` is neither a pass nor a
 * failure — it is "nobody has run this yet", and offering it would put a
 * candidate in a comparison on the strength of its archive being
 * readable.
 *
 * The API refuses such a stack anyway; this is the same rule stated
 * where the button lives, so the answer arrives before the click rather
 * than as a 422 after it.
 */
export function isSelectable(bundle: PluginBundleSummary): boolean {
  return bundle.status === "active" && bundle.validation_status === "loaded";
}

/** The single sentence a row shows about why it cannot be picked.
 *
 * Ordered by what the reader can act on: a disabled bundle is one
 * decision away from running, a failed one needs its code changed, and
 * an unverified one needs whatever stopped the run. Returns `null` when
 * there is nothing to say, so a caller cannot render an empty warning.
 */
export function blockedReason(bundle: PluginBundleSummary): string | null {
  if (isSelectable(bundle)) return null;
  if (bundle.status !== "active") return "disabled";
  if (bundle.validation_status === "failed") return "failed";
  return "unverified";
}

/** What to call each bundle's state, in one word a reader can act on.
 *
 * A projection, not a stored field. The server keeps three orthogonal
 * facts — may it be picked, did it load, is it published — because each
 * is set by a different act, and a stored label collapsing them would be
 * a fourth fact free to disagree with the other three. Collapsing them
 * *here*, for display, is what a status chip is for.
 *
 * Labels the whole list at once because one of the seven answers cannot
 * be reached from a single row: `superseded` means *another* revision of
 * the same `plugin_id` is the published one, so the verdict needs the
 * siblings. `awaiting` — nobody has published any of them — reads the
 * same to a picker but not to a reviewer, who has something to do about
 * exactly one of the two.
 */
export type BundleState =
  | "disabled"
  | "held"
  | "broken"
  | "checking"
  | "published"
  | "superseded"
  | "awaiting";

export function bundleStates(
  bundles: readonly PluginBundleSummary[],
  publishedIds: readonly string[],
): Map<string, BundleState> {
  const published = new Set(publishedIds);
  const pluginsWithOne = new Set(
    bundles.filter((bundle) => published.has(bundle.id)).map((bundle) => bundle.plugin_id),
  );
  return new Map(
    bundles.map((bundle) => {
      let state: BundleState;
      if (bundle.status === "disabled") state = "disabled";
      else if (bundle.status === "held") state = "held";
      else if (bundle.validation_status === "failed") state = "broken";
      else if (bundle.validation_status !== "loaded") state = "checking";
      else if (published.has(bundle.id)) state = "published";
      else state = pluginsWithOne.has(bundle.plugin_id) ? "superseded" : "awaiting";
      return [bundle.id, state] as const;
    }),
  );
}

export function publishedBundleIds(): Promise<string[]> {
  return authFetch<string[]>("/algorithms/plugins/published");
}

/** The stack id a benchmark would name this plugin by.
 *
 * `<global>+<plugin id>` — the plugin's own id rather than its display
 * name, because that is what the candidate hashes on. Shown in the row
 * so somebody reading a report can match the two.
 */
export function stackIdFor(bundle: PluginBundleSummary, globalPlanner = "astar"): string {
  return `${globalPlanner}+${bundle.plugin_id}`;
}

File: apps/api/planbench_api/routers/plugins.py
"""Imported algorithm bundles.

Sits under `/algorithms` beside the built-in catalogue, because from a
caller's point of view an imported algorithm and a registered one are
the same kind of thing — something a benchmark can name. What differs is
how it got here, and that is what this router is.

Reading the *catalogue* needs only an account: seeing an imported
algorithm, and seeing exactly why it cannot run, is what anybody
comparing candidates needs. Reading the *manifest in full* — entry
point, file listing, conformance log — needs `algorithm.inspect`,
because those describe code rather than capability. Importing one runs
the uploader's code on this server, and publishing one puts it in front
of everybody; both are the reviewer package
(`docs/plugin_import_security.md` §5).

**The governance routes are behind a flag.** Publish, unpublish, hold
and disable answer 404 until `PLANBENCH_ALGORITHM_GOVERNANCE` is on,
and that is deliberate rather than cautious: a kill switch is only safe
once the things downstream of it exist — a run that pinned which
revision it used, a queued job that stops when its bundle is turned off,
and an approval that can say its algorithm was withdrawn. Shipping the
switch first would give an operator a button whose consequences nothing
else in the system understands yet.
"""

from __future__ import annotations

import logging
from typing import Annotated

from fastapi import APIRouter, Depends, File, Form, HTTPException, Request, UploadFile, status
from pydantic import BaseModel

from planbench_api.accounts import Capability, User
from planbench_api.auth import ActiveUser, require_capability
from planbench_api.dependencies import get_plugin_service
from planbench_api.model_storage import CHUNK
from planbench_api.plugin_registry import PluginBundleSummary
from planbench_api.plugin_service import HostCompatibility, PluginBundleService

logger = logging.getLogger("planbench.api.plugins")

router = APIRouter(prefix="/algorithms/plugins", tags=["algorithms"])

Plugins = Annotated[PluginBundleService, Depends(get_plugin_service)]


InspectingUser = Annotated[User, Depends(require_capability(Capability.ALGORITHM_INSPECT))]


class PublicationView(BaseModel):
    """One row of the publication history, as the detail page shows it."""

    bundle_id: str
    revision: int
    published_at: str
    published_by_user_id: str | None = None
    superseded_at: str | None = None
    unpublished_at: str | None = None
    reason: str = ""
    is_current: bool


class PluginEventView(BaseModel):
    sequence: int
    revision: int
    actor_user_id: str | None
    actor_roles: str
    authorized_capability: str
    action: str
    reason: str
    created_at: str


class GovernanceRequest(BaseModel):
    reason: str = ""


def _require_governance(request: Request) -> None:
    """404 while the flag is off, because the route is not there yet.

    A 403 would say "you may not"; the truth is "this deployment has not
    turned publishing on", and a client that can tell the difference can
    hide the button rather than offer one that always fails.
    """
    if not request.app.state.deployment.algorithm_governance:
        raise HTTPException(
            status_code=404,
            detail=(
                "algorithm governance is not enabled on this deployment "
                "(PLANBENCH_ALGORITHM_GOVERNANCE)"
            ),
        )


class PluginBundleDetail(BaseModel):
    """One imported algorithm, with the verdict recomputed now.

    The compatibility report is not stored alongside the bundle and not
    cached here: a deployment can gain or lose a provider, and a verdict
    remembered from upload time would be a claim about a host that has
    since changed.
    """

    bundle: PluginBundleSummary
    compatibility: HostCompatibility
    #: The revision an engineer would actually get, or ``None`` while
    #: nobody has published one. Present for every reader: "why is this
    #: not in my picker?" is a question the person picking has to be able
    #: to answer without asking a reviewer.
    published_revision: int | None = None
    #: Manifest, entry point and history — code, not capability, so this
    #: half is empty for a reader without ``algorithm.inspect``.
    manifest: dict | None = None
    entry_point: str | None = None
    publications: list[PublicationView] | None = None


class PluginUpdateRequest(BaseModel):
    name: str | None = None
    description: str | None = None
    version: str | None = None
    status: str | None = None
    robot_profile_id: str | None = None


def _chunks(upload: UploadFile):
    """Stream the upload rather than reading it whole."""
    while True:
        block = upload.file.read(CHUNK)
        if not block:
            return
        yield block


@router.get("", response_model=list[PluginBundleSummary])
def list_plugins(plugins: Plugins, user: ActiveUser) -> list[PluginBundleSummary]:
    return [
        PluginBundleSummary.of(record, user.id, inspect=user.can(Capability.ALGORITHM_INSPECT))
        for record in plugins.list()
    ]


@router.post("", response_model=PluginBundleSummary, status_code=status.HTTP_201_CREATED)
async def import_plugin(
    plugins: Plugins,
    user: ActiveUser,
    name: Annotated[str, Form()],
    robot_profile_id: Annotated[str, Form()],
    version: Annotated[str, Form()] = "1",
    description: Annotated[str, Form()] = "",
    bundle: Annotated[UploadFile, File()] = ...,  # noqa: B008 - FastAPI form field
) -> PluginBundleSummary:
    """Import an algorithm bundle.

    `bundle` is a `.zip` of the directory holding the planner and its
    `.planbench-plugin/plugin.json`. Nothing in it is extracted or
    imported here: the archive's table of contents is read and its
    manifest is parsed, which is metadata parsing rather than execution.
    """
    record = plugins.upload(
        owner=user,
        name=name,
        version=version,
        description=description,
        robot_profile_id=robot_profile_id,
        filename=bundle.filename or "bundle.zip",
        chunks=_chunks(bundle),
    )
    return PluginBundleSummary.of(record, user.id, inspect=user.can(Capability.ALGORITHM_INSPECT))


@router.get("/{bundle_id}", response_model=PluginBundleDetail)
def get_plugin(bundle_id: str, plugins: Plugins, user: ActiveUser) -> PluginBundleDetail:
    """One bundle. How much of it comes back depends on the caller.

    The split is between *capability* and *code*. What an algorithm
    needs, what it supports and whether this host can run it are what an
    engineer uses to decide whether to pick it — withhold those and they
    cannot configure a candidate. Its entry point, its file listing and
    its conformance log describe the code itself, and reading code is
    the reviewer's job.
    """
    record = plugins.get(bundle_id)
    inspecting = user.can(Capability.ALGORITHM_INSPECT)
    current = plugins.publication(record.plugin_id)
    return PluginBundleDetail(
        bundle=PluginBundleSummary.of(record, user.id, inspect=inspecting),
        compatibility=plugins.compatibility(bundle_id),
        published_revision=current.revision if current is not None else None,
        manifest=record.manifest if inspecting else None,
        entry_point=record.entry_point if inspecting else None,
        publications=(
            [_publication_view(row) for row in plugins.publication_history(record.plugin_id)]
            if inspecting
            else None
        ),
    )


def _publication_view(row) -> PublicationView:
    return PublicationView(
        bundle_id=row.bundle_id,
        revision=row.revision,
        published_at=row.published_at,
        published_by_user_id=row.published_by_user_id,
        superseded_at=row.superseded_at,
        unpublished_at=row.unpublished_at,
        reason=row.reason,
        is_current=row.is_current,
    )


@router.get("/{bundle_id}/events", response_model=list[PluginEventView])
def plugin_events(bundle_id: str, plugins: Plugins, _: InspectingUser) -> list[PluginEventView]:
    """Who did what to this bundle, oldest first."""
    return [
        PluginEventView(
            sequence=event.sequence,
            revision=event.revision,
            actor_user_id=event.actor_user_id,
            actor_roles=event.actor_roles,
            authorized_capability=event.authorized_capability,
            action=event.action,
            reason=event.reason,
            created_at=event.created_at,
        )
        for event in plugins.events(bundle_id)
    ]


@router.post("/{bundle_id}/publish", response_model=PluginBundleSummary)
def publish_plugin(
    bundle_id: str,
    payload: GovernanceRequest,
    plugins: Plugins,
    user: ActiveUser,
    request: Request,
) -> PluginBundleSummary:
    """Make this the revision every engineer gets."""
    _require_governance(request)
    return PluginBundleSummary.of(
        plugins.publish(bundle_id, user, payload.reason),
        user.id,
        inspect=user.can(Capability.ALGORITHM_INSPECT),
    )


@router.post("/{bundle_id}/unpublish", response_model=PluginBundleSummary)
def unpublish_plugin(
    bundle_id: str,
    payload: GovernanceRequest,
    plugins: Plugins,
    user: ActiveUser,
    request: Request,
) -> PluginBundleSummary:
    """Take it back out. Publishing again restores it."""
    _require_governance(request)
    return PluginBundleSummary.of(
        plugins.unpublish(bundle_id, user, payload.reason),
        user.id,
        inspect=user.can(Capability.ALGORITHM_INSPECT),
    )


@router.post("/{bundle_id}/hold", response_model=PluginBundleSummary)
def hold_plugin(
    bundle_id: str,
    payload: GovernanceRequest,
    plugins: Plugins,
    user: ActiveUser,
    request: Request,
) -> PluginBundleSummary:
    _require_governance(request)
    return PluginBundleSummary.of(
        plugins.hold(bundle_id, user, payload.reason),
        user.id,
        inspect=user.can(Capability.ALGORITHM_INSPECT),
    )


@router.post("/{bundle_id}/release-hold", response_model=PluginBundleSummary)
def release_plugin_hold(
    bundle_id: str,
    payload: GovernanceRequest,
    plugins: Plugins,
    user: ActiveUser,
    request: Request,
) -> PluginBundleSummary:
    _require_governance(request)
    return PluginBundleSummary.of(
        plugins.release_hold(bundle_id, user, payload.reason),
        user.id,
        inspect=user.can(Capability.ALGORITHM_INSPECT),
    )


@router.post("/{bundle_id}/disable", response_model=PluginBundleSummary)
def disable_plugin(
    bundle_id: str,
    payload: GovernanceRequest,
    plugins: Plugins,
    user: ActiveUser,
    request: Request,
) -> PluginBundleSummary:
    """Retire it for good, on governance grounds. Needs a reason.

    The administrator's kill switch is a different route with the same
    effect, so the audit row can say which job the caller was doing —
    retiring an algorithm and responding to an incident are not the same
    act, and a shared endpoint would have to guess.
    """
    _require_governance(request)
    record = plugins.disable(
        bundle_id, user, payload.reason, capability=Capability.ALGORITHM_DISABLE
    )
    _note_approvals_that_rested_on_it(request, record, user, payload.reason)
    return PluginBundleSummary.of(record, user.id, inspect=user.can(Capability.ALGORITHM_INSPECT))


def _note_approvals_that_rested_on_it(request: Request, record, user, reason: str) -> None:
    """Write into each affected run's journal that its algorithm went away.

    **Not a withdrawal.** The system does not sign anything on a
    reviewer's behalf, and disabling can happen for a security fix, a
    crash, a dependency, or an investigation — none of which prove the
    original recommendation was wrong. What it does is leave a dated
    entry beside the approval, so somebody reading that journal later
    finds out *there* rather than by noticing the algorithm is gone.

    Failure here is logged and swallowed: an incomplete journal entry is
    bad, and an algorithm that could not be turned off because writing
    one failed is worse.
    """
    runs = request.app.state.repos.decision_runs
    try:
        affected = [
            run
            for run in runs.list()
            if run.config_state == "approved"
            and any(
                entry.get("bundle_id") == record.id
                for entry in (getattr(run, "candidates", []) or [])
            )
        ]
    except Exception:  # noqa: BLE001 - see the docstring
        logger.warning("could not look for approvals resting on this bundle", exc_info=True)
        return
    for run in affected:
        try:
            runs.append_event(
                run.id,
                "algorithm_disabled_after_approval",
                user.id,
                user.nickname,
                run.config_state,
                run.config_state,
                f"{record.label} (revision {record.revision}) was disabled: {reason}",
            )
        except Exception:  # noqa: BLE001 - see the docstring
            logger.warning(
                "could not record the disable against a run", extra={"context": {"run": run.id}}
            )


@router.post("/{bundle_id}/validate", response_model=PluginBundleSummary)
def validate_plugin(bundle_id: str, plugins: Plugins, user: ActiveUser) -> PluginBundleSummary:
    """Unpack the bundle and run the conformance suite again.

    Runs the uploader's code, so it is the import privilege rather than
    the read one. Worth re-asking after a deployment changes: a bundle
    left unchecked because a provider was missing is not a bundle that
    failed.
    """
    return PluginBundleSummary.of(
        plugins.revalidate(bundle_id, user), user.id, inspect=user.can(Capability.ALGORITHM_INSPECT)
    )


@router.patch("/{bundle_id}", response_model=PluginBundleSummary)
def update_plugin(
    bundle_id: str, payload: PluginUpdateRequest, plugins: Plugins, user: ActiveUser
) -> PluginBundleSummary:
    changes = {key: value for key, value in payload.model_dump().items() if value is not None}
    return PluginBundleSummary.of(
        plugins.update(bundle_id, changes, user),
        user.id,
        inspect=user.can(Capability.ALGORITHM_INSPECT),
    )
